File: server/EuroPad.Server/Protocol/InputFrame.cs
namespace EuroPad.Server.Protocol;

[Flags]
public enum ButtonsLo : ushort
{
    None = 0,
    DpadUp = 1 << 0,
    DpadRight = 1 << 1,
    DpadDown = 1 << 2,
    DpadLeft = 1 << 3,
    Start = 1 << 4,
    Back = 1 << 5,
    Lb = 1 << 6,
    Rb = 1 << 7,
    A = 1 << 8,
    B = 1 << 9,
    X = 1 << 10,
    Y = 1 << 11,
    Guide = 1 << 12,
}

[Flags]
public enum ButtonsHi : ushort
{
    None = 0,
    IndL = 1 << 0,
    IndR = 1 << 1,
    Hazard = 1 << 2,
    Horn = 1 << 3,
    AirHorn = 1 << 4,
    Handbrake = 1 << 5,
    Lights = 1 << 6,
    Beam = 1 << 7,
    Warning = 1 << 8,
    Wipers = 1 << 9,
    ExhBrake = 1 << 10,
    DiffLock = 1 << 11,
    AxleRaise = 1 << 12,
    Engine = 1 << 13,
    GearUp = 1 << 14,
    GearDn = 1 << 15,
}

public static class AxisIndex
{
    public const int Lx = 0;
    public const int Ly = 1;
    public const int Rx = 2;
    public const int Ry = 3;
    public const int Lt = 4;
    public const int Rt = 5;
    public const int Steer = 6;
    public const int Aux0 = 7;
}

public readonly struct InputFrame
{
    public readonly byte Flags;
    public readonly ushort Seq;
    public readonly uint TimestampMs;
    public readonly ushort ButtonsLoRaw;
    public readonly ushort ButtonsHiRaw;
    public readonly short[] Axes;

    public InputFrame(byte flags, ushort seq, uint timestampMs, ushort buttonsLo, ushort buttonsHi, short[] axes)
    {
        Flags = flags;
        Seq = seq;
        TimestampMs = timestampMs;
        ButtonsLoRaw = buttonsLo;
        ButtonsHiRaw = buttonsHi;
        Axes = axes;
    }

    public bool IsPingRequest => (Flags & Proto.FlagPingRequest) != 0;

    public bool IsHello => (Flags & Proto.FlagHello) != 0;

    public bool IsConfig => (Flags & Proto.FlagConfig) != 0;

    /// <summary>
    /// Config-frame payload: ButtonsHi[0] = char count, ButtonsHi[1] + the 16 axis bytes carry the
    /// ASCII characters (little-endian order). Max 17 chars — key names fit with room to spare.
    /// </summary>
    public string PayloadText
    {
        get
        {
            byte len = (byte)(ButtonsHiRaw & 0xFF);
            if (len == 0 || len > 17) return string.Empty;
            var chars = new System.Text.StringBuilder(len);
            void Add(byte b)
            {
                if (b >= 0x20 && chars.Length < len) chars.Append((char)b);
            }
            Add((byte)(ButtonsHiRaw >> 8));
            foreach (var axis in Axes)
            {
                Add((byte)(axis & 0xFF));
                Add((byte)((axis >> 8) & 0xFF));
            }
            return chars.ToString();
        }
    }

    public int Pin => ButtonsLoRaw;

    public int Axis(int i) => (uint)i < (uint)Axes.Length ? Axes[i] : 0;
}

File: app/app/src/main/java/com/europad/app/input/InputFrame.kt
package com.europad.app.input

class InputFrame {
    var buttonsLo: Int = 0
    var buttonsHi: Int = 0
    val axes = ShortArray(8)

    fun reset() {
        buttonsLo = 0
        buttonsHi = 0
        axes.fill(0)
    }

    fun copyFrom(other: InputFrame) {
        buttonsLo = other.buttonsLo
        buttonsHi = other.buttonsHi
        System.arraycopy(other.axes, 0, axes, 0, axes.size)
    }
}

object ButtonLo {
    const val DPAD_UP = 1 shl 0
    const val DPAD_RIGHT = 1 shl 1
    const val DPAD_DOWN = 1 shl 2
    const val DPAD_LEFT = 1 shl 3
    const val START = 1 shl 4
    const val BACK = 1 shl 5
    const val LB = 1 shl 6
    const val RB = 1 shl 7
    const val A = 1 shl 8
    const val B = 1 shl 9
    const val X = 1 shl 10
    const val Y = 1 shl 11
    const val GUIDE = 1 shl 12
}

object ButtonHi {
    const val IND_L = 1 shl 0
    const val IND_R = 1 shl 1
    const val HAZARD = 1 shl 2
    const val HORN = 1 shl 3
    const val AIR_HORN = 1 shl 4
    const val HANDBRAKE = 1 shl 5
    const val LIGHTS = 1 shl 6
    const val BEAM = 1 shl 7
    const val WARNING = 1 shl 8
    const val WIPERS = 1 shl 9
    const val EXH_BRAKE = 1 shl 10
    const val DIFF_LOCK = 1 shl 11
    const val AXLE_RAISE = 1 shl 12
    const val ENGINE = 1 shl 13
    const val GEAR_UP = 1 shl 14
    const val GEAR_DN = 1 shl 15
}
